native/dlasy2.go

```diff
@@ -0,0 +1,270 @@
+		// Set pivots less than smin to smin.
+		bi := blas64.Implementation()
+		ipiv := bi.Idamax(len(tmp), tmp[:], 1)
+		u11 := tmp[ipiv]
+		if math.Abs(u11) <= smin {
+			ok = false
+			u11 = smin
+		}
+		locu12 := [4]int{2, 3, 0, 1}
+		u12 := tmp[locu12[ipiv]]
+		locl21 := [4]int{1, 0, 3, 2}
+		l21 := tmp[locl21[ipiv]] / u11
+		locu22 := [4]int{3, 2, 1, 0}
+		u22 := tmp[locu22[ipiv]] - u12*l21
+		if math.Abs(u22) <= smin {
+			ok = false
+			u22 = smin
+		}
+		if ipiv&0x1 != 0 { // true for ipiv equal to 1 and 3.
+			btmp[0], btmp[1] = btmp[1], btmp[0]-l21*btmp[1]
+		} else {
+			btmp[1] -= l21 * btmp[0]
+		}
+		scale = 1
+		if 2*smlnum*math.Abs(btmp[1]) > math.Abs(u22) || 2*smlnum*math.Abs(btmp[0]) > math.Abs(u11) {
+			scale = 0.5 / math.Max(math.Abs(btmp[0]), math.Abs(btmp[1]))
+			btmp[0] *= scale
+			btmp[1] *= scale
+		}
+		x21 := btmp[1] / u22
+		x20 := btmp[0]/u11 - (u12/u11)*x21
+		if ipiv&0x2 != 0 { // true for ipiv equal to 2 and 3.
+			x20, x21 = x21, x20
+		}
+		x[0] = x20
+		if n1 == 1 {
+			x[1] = x21
+			xnorm = math.Abs(x[0]) + math.Abs(x[1])
+		} else {
+			x[ldx] = x21
+			xnorm = math.Max(math.Abs(x[0]), math.Abs(x[ldx]))
+		}
+		return scale, xnorm, ok
+	}
+
+	// 2×2 case: op[TL11 TL12]*[X11 X12] + SGN*[X11 X12]*op[TR11 TR12] = [B11 B12].
+	//             [TL21 TL22] [X21 X22]       [X21 X22]   [TR21 TR22]   [B21 B22]
+	//
+	// Solve equivalent 4×4 system using complete pivoting.
+	// Set pivots less than smin to smin.
+
+	smin := math.Max(math.Abs(tr[0]), math.Abs(tr[1]))
+	smin = math.Max(smin, math.Max(math.Abs(tr[ldtr]), math.Abs(tr[ldtr+1])))
+	smin = math.Max(smin, math.Max(math.Abs(tl[0]), math.Abs(tl[1])))
+	smin = math.Max(smin, math.Max(math.Abs(tl[ldtl]), math.Abs(tl[ldtl+1])))
+	smin = math.Max(eps*smin, smlnum)
+
+	var t16 [16]float64 // t16 is used as a 4×4 column-major matrix.
+	t16[0*4+0] = tl[0] + sgn*tr[0]
+	t16[1*4+1] = tl[ldtl+1] + sgn*tr[0]
+	t16[2*4+2] = tl[0] + sgn*tr[ldtr+1]
+	t16[3*4+3] = tl[ldtl+1] + sgn*tr[ldtr+1]
+	if tranl {
+		t16[0*4+1] = tl[ldtl]
+		t16[1*4+0] = tl[1]
+		t16[2*4+3] = tl[ldtl]
+		t16[3*4+2] = tl[1]
+	} else {
+		t16[0*4+1] = tl[1]
+		t16[1*4+0] = tl[ldtl]
+		t16[2*4+3] = tl[1]
+		t16[3*4+2] = tl[ldtl]
+	}
+	if tranr {
+		t16[0*4+2] = sgn * tr[1]
+		t16[1*4+3] = sgn * tr[1]
+		t16[2*4+0] = sgn * tr[ldtr]
+		t16[3*4+1] = sgn * tr[ldtr]
+	} else {
+		t16[0*4+2] = sgn * tr[ldtr]
+		t16[1*4+3] = sgn * tr[ldtr]
+		t16[2*4+0] = sgn * tr[1]
+		t16[3*4+1] = sgn * tr[1]
+	}
+
+	var btmp [4]float64
+	btmp[0] = b[0]
+	btmp[1] = b[ldb]
+	btmp[2] = b[1]
+	btmp[3] = b[ldb+1]
+
+	// Perform elimination.
+	bi := blas64.Implementation()
+	var jpiv [4]int
+	for i := 0; i < 3; i++ {
+		var (
+			xmax       float64
+			ipsv, jpsv int
+		)
+		for ip := i; ip < 4; ip++ {
+			for jp := i; jp < 4; jp++ {
+				if math.Abs(t16[ip*4+jp]) >= xmax {
+					xmax = math.Abs(t16[ip*4+jp])
+					ipsv = ip
+					jpsv = jp
+				}
+			}
+		}
+		if ipsv != i {
+			bi.Dswap(4, t16[ipsv*4:], 1, t16[i*4:], 1)
```

Comment:
I think I misled you because of my (not unusual) momentary confusion, sorry. tmp and t16 are both row-major. Otherwise this would not work (Dswap just swaps vectors, does not care about row/col major), also at other places the indexing is row-major. I got confused by how they map b to btmp and tmp to x.

With this misconception cleared, maybe I can use [2][2]tmp and [4][4]t16 (renamed to just t?) after all if I replaced these calls to blas with loops and assignments.